File: book/fridgets/f#/Fridgets/FrTranslate.fs
namespace Fridgets

open System.Windows
open System.Windows.Media
open Sodium.Frp

type FrTranslate =
    private {
        reify : Cell<Size option> -> Stream<MouseEvent> -> Stream<KeyEvent> -> Cell<int64> -> Supply.T -> Output
    }
    interface IFridget with
        member this.Reify size sMouse sKey focus idSupply = this.reify size sMouse sKey focus idSupply

[<CompilationRepresentationAttribute(CompilationRepresentationFlags.ModuleSuffix)>]
module FrTranslate =
    let create fridget offset =
        let reify size sMouse sKey focus idSupply =
            
            let offsetMouseEvent sMouse (offset : Point) =
                let getPosition () =
                    let p = sMouse.getPosition ()
                    Point(p.X - offset.X, p.Y - offset.Y)
                { args = sMouse.args; getPosition = getPosition }
            let sMouse = sMouse |> snapshotC offset offsetMouseEvent

            let fo = Fridget.reify fridget size sMouse sKey focus idSupply

            let offsetDrawable drawable (offset : Point) =
                DrawableDelegate (fun d ->
                    d.PushTransform(TranslateTransform(offset.X, offset.Y))
                    drawable |> DrawableDelegate.invoke d
                    d.Pop())

            {
                drawable = (fo.drawable, offset) |> lift2C offsetDrawable
                desiredSize = fo.desiredSize
                sChangeFocus = fo.sChangeFocus
            }
        
        {
            reify = reify
        }

File: book/fridgets/c#/Fridgets/FrTranslate.cs
using System.Windows;
using System.Windows.Media;
using Sodium.Frp;

namespace Fridgets
{
    public class FrTranslate : Fridget
    {
        public FrTranslate(Fridget fr, Cell<Point> offset)
            : base((size, sMouse, sKey, focus, idSupply) =>
            {
                Stream<MouseEvent> sMouseNew =
                    sMouse.Snapshot(offset, (e, o) =>
                        new MouseEvent(e.Args, () =>
                        {
                            Point p = e.GetPosition();
                            return new Point(p.X - o.X, p.Y - o.Y);
                        }));
                Output fo = fr.Reify(size, sMouseNew, sKey, focus, idSupply);
                Cell<DrawableDelegate> drawableNew = fo.Drawable.Lift(offset,
                    (dr, o) => new DrawableDelegate(d =>
                    {
                        d.PushTransform(new TranslateTransform(o.X, o.Y));
                        dr(d);
                        d.Pop();
                    }));
                return new Output(drawableNew, fo.DesiredSize, fo.SChangeFocus);
            })
        {
        }
    }
}